domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/ApplicationLifecycleListener.kt

```diff
@@ -0,0 +1,10 @@
+package org.oppia.android.domain.oppialogger.analytics
+
+/** Listener for when the app goes to the foreground or the background. */
+interface ApplicationLifecycleListener {
```

Comment:
One implementation of these listeners that I attempted was to implement it in `ApplicationLifecycleObserver`, similar to [ApplicationStartupListener](https://github.com/oppia/oppia-android/blob/1ba98ddb66590e7879ec01928ac6e59cd9da4408/domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/ApplicationLifecycleObserver.kt#L80). The only reason it does not work is that an implementation to start and stop a timer here will be triggered even when the user is not currently in the exploration screen.

I am not certain how best to enforce a check for current screen when the app is in foreground, to ensure the timer is only started when in ExplorationActivity.

I think the alternative would be to use the activity's lifecycle onPause/onResume. What do you think about this.

Reply:
As discussed in meeting today, I suggest refactoring the existing ApplicationLifecycleObserver such that we have a general handler for forwarding these lifecycle events, e.g.:

```kotlin
class ApplicationLifecycleObserver @Inject constructor(
  private val applicationLifecycleListeners: Set<ApplicationLifecycleListener>
): ApplicationStartupListener, LifecycleObserver {
  override fun onCreate() {
    ProcessLifecycleOwner.get().lifecycle.addObserver(this)
  }

  @OnLifecycleEvent(Lifecycle.Event.ON_START)
  fun onAppInForeground() {
    applicationLifecycleListeners.forEach(ApplicationLifecycleListener::onAppInForeground)
  }

  /** Occurs when application goes to background. */
  @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
  fun onAppInBackground() {
    applicationLifecycleListeners.forEach(ApplicationLifecycleListener::onAppInBackground)
  }
}
```

Reply:
I did this refactor and implemented the listeners in the `ExplorationProgressListenerImpl` as discussed, but I encontered these situations:
1. DRY: if the callbacks from both interfaces do the same thing, can we consolidate their implementations?
2. Not sure how the onAppInBackground() callback will receive the `profileId` and `topicId` needed when calling `setExplorationSessionPaused(profileId, topicId)` the callback itself takes no arguments, which is ideal since we fire it in the ApplicationLifecycleObserver like in your example above.

Here is what the impl looks like now:

```
@Singleton
class ExplorationProgressListenerImpl @Inject constructor(
  private val explorationActiveTimeController: ExplorationActiveTimeController
) : ExplorationProgressListener, ApplicationLifecycleListener {
  override fun onExplorationSessionStarted() {
    explorationActiveTimeController.setExplorationSessionStarted()
  }

  override fun onExplorationSessionPaused(profileId: ProfileId, topicId: String) {
    explorationActiveTimeController.setExplorationSessionPaused(profileId, topicId)
  }

  override fun onAppInForeground() {
    explorationActiveTimeController.setExplorationSessionStarted()
  }

  override fun onAppInBackground() {
    print("app is backgrounded")
  }
}
```
This way, with some logging, I can see that  `explorationActiveTimeController.setExplorationSessionStarted() `is always called, as soon as the app  comes into the foreground, regardless of the screen.

3. Where can I move this implementation so that the background/foreground behaviour is only reacted to within an exploration context?

Reply:
I think (2) and (3) relate to another comment I left in the listener impl--the correct way to do this is to probably use the signals of session activity & app activity independently to determine the current session state and use that for decision making. I think this also addresses (1) since I see the two listeners as providers of completely distinct information that we're using together to determine a particular domain state from which we can make decisions.